pkg/template/template.go

```diff
@@ -19,37 +20,621 @@ package template
+// {{- "Hi There" | saveAs "TaskResult.msg" .Values | noop -}}
+// {{- .Values.TaskResult.msg -}}
+//
+// Above will print 'Hi There'
+//
+// Assumption here is .Values is of type 'map[string]interface{}'
+func saveIf(fields string, values map[string]interface{}, value interface{}) interface{} {
+	fieldsArr := strings.Split(fields, ".")
+	oldValue := util.GetNestedField(values, fieldsArr...)
+	// will not override the old value
+	if !empty(oldValue) {
+		return oldValue
+	}
+
+	util.SetNestedField(values, value, fieldsArr...)
+	return value
+}
+
+// notFoundErr throws NotFoundError if given object is empty
+//
+// NOTE:
+//  This function is intended to be used as a go template function
+//
+// Example:
+// {{- "" | notFoundErr "empty object" | saveIf "errMsg" .Values | noop -}}
+//
+// Above returns NotFoundError during template execution. However this
+// does not result in a runtime error.
+//
+// {{- "I am not empty" | notFoundErr "empty object" | saveIf "errMsg" .Values | noop -}}
+//
+// Above does not return any error during template execution.
+//
+// Assumption here is .Values is of type 'map[string]interface{}'
+func notFoundErr(errMessage string, given interface{}) (err error) {
+	if !empty(given) {
+		// no error if not empty
+		return
+	}
+
+	if len(errMessage) == 0 {
+		errMessage = "item is not found"
+	}
+
+	err = &NotFoundError{
+		err: errMessage,
+	}
+
+	return
+}
+
+// verifyErr returns VerifyError if given verification flag failed i.e. is true
+//
+// NOTE:
+//  This is intended to be used as a go template function
+//
+// Example:
+// {{- "" | empty | verifyErr "empty value provided" | saveIf "errMsg" .Values | noop -}}
+//
+// Above returns VerifyError during template execution. However this
+// does not result in a runtime error.
+//
+// {{- "I am not empty" | empty | verifyErr "empty value provided" | saveIf "errMsg" .Values | noop -}}
+//
+// Above does not return any error during template execution.
+//
+// Assumption here is .Values is of type 'map[string]interface{}'
+func verifyErr(errMessage string, hasVerificationFailed bool) (err error) {
+	if !hasVerificationFailed {
+		// no error if verification did pass successfully
+		return
+	}
+
+	if len(errMessage) == 0 {
+		errMessage = "verification failed"
+	}
+
+	err = &VerifyError{
+		err: errMessage,
+	}
+
+	return
+}
+
+// toYaml takes an interface, marshals it to yaml, and returns a string. It will
+// always return a string, even on marshal error (empty string).
+//
+// This is designed to be called from a template.
 //
 // NOTE: Borrowed from a similar function in helm
 //  https://github.com/kubernetes/helm/blob/master/pkg/chartutil/files.go
-func toYaml(yml string) string {
-	var m interface{}
-	err := yaml.Unmarshal([]byte(yml), &m)
+func toYaml(v interface{}) (yamlstr string) {
+	data, err := yaml.Marshal(v)
 	if err != nil {
-		// Swallow errors inside of a template.
-		return ""
+		// error is swallowed
```

Comment:
can u pls explain why the error is swallowed? If we return the error, we can get some info at which line this yaml marshal has failed.

Reply:
in templating it is advised to handle the errors than letting them bubble out as runtime errors. I will try to manage these errors in a way so that it does not bubble out as un-handled error.